object-handlers.go

```diff
@@ -594,51 +594,78 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 		// Create anonymous object.
 		md5Sum, err = api.ObjectAPI.PutObject(bucket, object, size, r.Body, metadata)
 	case authTypePresigned, authTypeSigned:
-		// Initialize a pipe for data pipe line.
-		reader, writer := io.Pipe()
-		var wg = &sync.WaitGroup{}
-		// Start writing in a routine.
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			shaWriter := sha256.New()
-			multiWriter := io.MultiWriter(shaWriter, writer)
-			if _, wErr := io.CopyN(multiWriter, r.Body, size); wErr != nil {
-				// Pipe closed.
-				if wErr == io.ErrClosedPipe {
-					return
-				}
-				errorIf(wErr, "Unable to read from HTTP body.")
-				writer.CloseWithError(wErr)
-				return
-			}
-			shaPayload := shaWriter.Sum(nil)
-			validateRegion := true // Validate region.
+		shaHeader := r.Header.Get("x-amz-content-sha256")
+		validateRegion := true // Validate region.
+
+		// http Header "x-amz-content-sha256" == "UNSIGNED-PAYLOAD" indicates that the
+		// client did not calculate sha256 of the payload. Hence we skip calculating sha256.
+		// We also skip calculating sha256 for presigned requests without  "x-amz-content-sha256" header.
+		skipSha := shaHeader == "UNSIGNED-PAYLOAD" || (isRequestPresignedSignatureV4(r) && shaHeader == "")
```

Comment:
This looks like a recent addition - http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html 

Earlier developer forums indicate that 'UNSIGNED-PAYLOAD' only to be meant for preSigned requests.

https://forums.aws.amazon.com/message.jspa?messageID=573632
```
UNSIGNED-PAYLOAD can be used only with a query-string authentication. If you use Authorization header authentication, it cannot be used.
As an option, you can use chunked transfer, so will have to calculate hashes for small chunks of data than can be buffered for hashing.
Also, you can still use older Signature V2 , though it won't work with regions created after 30-jan-2014.
```

They must have relaxed it.  Can you reference that link here and move this a function of its own? in auth-handler.go as 'isRequestPayloadSigned()' ?